mir-ci/mir_ci/tests/test_mir_flutter_app.py

```diff
@@ -0,0 +1,439 @@
+${{BUTTON_TIP_0}}                             {ASSETS_PATH}/button_tip_0.png
+${{BUTTON_CLOSE_FOCUSED}}                     {ASSETS_PATH}/button_close_focused.png
+${{BUTTON_CUSTOM_PRESET}}                     {ASSETS_PATH}/button_custom_preset.png
+${{BUTTON_SET_DEFAULTS}}                      {ASSETS_PATH}/button_set_defaults.png
+${{BUTTON_APPLY}}                             {ASSETS_PATH}/button_apply.png
+
+${{DROPDOWN_PRESET_LABEL}}                    {ASSETS_PATH}/dropdown_preset_label.png
+${{DROPDOWN_OPTION_CUSTOM}}                   {ASSETS_PATH}/dropdown_option_custom.png
+${{DROPDOWN_OPTION_LEFT}}                     {ASSETS_PATH}/dropdown_option_left.png
+${{DROPDOWN_PARENT_ANCHOR_LABEL}}             {ASSETS_PATH}/dropdown_parent_anchor_label.png
+${{DROPDOWN_CHILD_ANCHOR_LABEL}}              {ASSETS_PATH}/dropdown_child_anchor_label.png
+
+${{GROUP_NEW_WINDOW}}                         {ASSETS_PATH}/group_new_window.png
+
+${{LABEL_CONSTRAINTS}}                        {ASSETS_PATH}/label_constraints.png
+
+${{WINDOW_MAIN}}                              {ASSETS_PATH}/window_main.png
+${{WINDOW_REGULAR_0_FOCUSED}}                 {ASSETS_PATH}/window_regular_0_focused.png
+${{WINDOW_FLOATING_REGULAR_0_FOCUSED}}        {ASSETS_PATH}/window_floating_regular_0_focused.png
+${{WINDOW_FLOATING_REGULAR_0_NON_FOCUSED}}    {ASSETS_PATH}/window_floating_regular_0_non_focused.png
+${{WINDOW_DIALOG_FOCUSED}}                    {ASSETS_PATH}/window_dialog_focused.png
+${{WINDOW_SATELLITE_1}}                       {ASSETS_PATH}/window_satellite_1.png
+${{WINDOW_POPUP_1}}                           {ASSETS_PATH}/window_popup_1.png
+${{WINDOW_TIP_1}}                             {ASSETS_PATH}/window_tip_1.png
+
+${{EXPECTED_SATELLITE_PLACEMENT}}             {ASSETS_PATH}/expected_satellite_placement.png
+${{EXPECTED_WINDOW_BEFORE_MOVE}}              {ASSETS_PATH}/expected_window_before_move.png
+${{EXPECTED_WINDOW_AFTER_MOVE}}               {ASSETS_PATH}/expected_window_after_move.png
+${{EXPECTED_POPUP_PLACEMENT_SLIDE}}           {ASSETS_PATH}/expected_popup_placement_slide.png
+${{EXPECTED_POPUP_PLACEMENT_FLIP}}            {ASSETS_PATH}/expected_popup_placement_flip.png
+${{EXPECTED_POPUP_PLACEMENT_RESIZE}}          {ASSETS_PATH}/expected_popup_placement_resize.png
+
+${{VERTICAL_DISTANCE_BETWEEN_OPTIONS}}        48
+"""
+
+ROBOT_KEYWORDS = """\
+Click And Sleep
+    Click LEFT Button
+    Sleep    0.5
+
+Close Focused Toplevel Window
+    Walk Pointer To ${BUTTON_CLOSE_FOCUSED}
+    Click And Sleep
+
+Close Dialog
+    ${pos}=    Walk Pointer To ${WINDOW_DIALOG_FOCUSED}
+    ${pos}=    Displace ${pos} By (112, 0)
+    Walk Pointer To ${pos}
+    Click And Sleep
+
+Select ${preset} Positioner Preset
+    # Reset first to make sure the dropdown list is positioned as expected
+    ${pos}=    Move Pointer To ${DROPDOWN_PRESET_LABEL}
+    ${pos}=    Displace ${pos} By (0, 20)
+    Walk Pointer To ${pos}
+    Click And Sleep
```

Comment:
Is this snippet used to "click off" of the dropdown in the event that it is already opened?

Reply:
Yes, but not only that. It also ensures that when the list of presets is opened, the widget will be displayed at the same vertical position, allowing each option to be selected by moving the pointer a fixed vertical displacement. I've attached an image below to illustrate this.:

![image](https://github.com/canonical/mir-ci/assets/43485228/088702f2-13eb-4acc-8f9d-9634cf0dedeb)

If the Custom preset is selected (as seen on the left), Flutter displays the list slightly towards the top of the main window. If any other option is selected, the list will be displayed as shown on the right. This snippet ensures this is always the case. From this configuration, we can select any option by moving the pointer in multiples of 48 pixels IIRC (e.g., 48 for "Custom", -48 for "Bottom Right", -96 for "Bottom", etc.).

Alternatively, I could have used image templates, but that would require more complex logic and a lot of (14?) image templates because each option can either be selected (gray background) or not selected (white background).

Reply:
I wonder if image templates wouldn't be as error prone here, but I will leave that up to you to decide. Doing pixel-walking always feels sketchy to me

Reply:
Yeah, I'd also prefer using templates instead. In this case it would help to have a keyword like `Match Any Of ${list_of_templates}` where `list_of_templates` could contain the images for the option with white and gray background. WDYT?

Reply:
Nice, I'll add this functionality then.

Reply:
Done.